samples/99_zil_interpreter/app/parser.rb

```diff
@@ -0,0 +1,375 @@
+    raise "Expression recursion overflow" if @expr_depth > 50
+
+    expr_char = peek
+    if expr_char == '<' #  Starts with '<' then read_form
+      log "+ FORM " if debug_flag
+      expr = read_form
+    elsif expr_char == '(' #  Starts with '(' then read_list
+      log "+ LIST " if debug_flag
+      expr = read_list
+    elsif expr_char == '"' #  Starts with '"' then read_string
+      log "+ STRING " if debug_flag
+      expr = read_string
+    elsif expr_char == ';' #  Starts with ';' then read_comment
+      log "+ COMMENT " if debug_flag
+      expr = read_comment
+    elsif expr_char == '#' && peek_string(5) == '#DECL'
+      log "+ DECL " if debug_flag
+      expr = read_decl
+    elsif expr_char == '%' && peek_string(6) == '%<COND'
+      log "+ COND MACRO " if debug_flag
+      expr = read_cond_macro
+    elsif expr_char == "'"
+      log "+ MACRO " if debug_flag
+      expr = read_macro
+    elsif is_fix?(expr_char) # Starts with 0123456789 then read_fix
+      log "+ FIX " if debug_flag
+      expr = read_fix
+    else
+      #  else read atom
+      log "+ ATOM " if debug_flag
+      expr = read_atom
+    end
+
+    @expr_depth -= 1
+    expr
+  end
+
+  def read_fix
+    # UC1: 123456
+    # UC2: *3777*
+    # UC3: #2 0111
+
+    fix = ""
+
+    # step 1: just load the string based on the three types
+    while @buf_index < @buffer.length - 1 # iterate characters in line
+      buf_char = @buffer[@buf_index]
+      @buf_index += 1
+
+      if FIX_CHARS.include?(buf_char)
+        #log "[TMP] " + buf_char
+        fix += buf_char
+      elsif buf_char == " " && fix == '#2'
+        fix += buf_char
+      else
+        @buf_index -= 1
+        break
+      end
+    end
+
+    # step 2: convert to number
+    if fix.start_with?('*')
+      fix = fix.gsub('*', '')
+      fix = fix.to_i(base=8)
+    elsif fix.start_with?('#2')
+      fix = fix.gsub('#2 ', '')
+      fix = fix.to_i(base=2)
+    else
+      fix = fix.to_i
+    end
+
+    fix
+  end
+
+  # <...>
+  def read_form
+    lt = read_char # 1) Read '<'
+    atom = read_atom # 2) Read atom
+    skip_whitespace # 3) Skip whitespace
+
+    # 4) Read params (form, list, string, fix)
+    expr = []
+    last_buf_index = -1 # make sure we don't get into an endless loop
+    done = false
+    until done
+      if @buf_index == last_buf_index # endless loop detected
+        log "ENDLESS at '" + peek + "'"
+        raise "ENDLESS"
+      end
+
+      last_buf_index = @buf_index
+
+      ft = peek
+      if ft == '>' # handle special case of empty form
+        done = true
+      else
+        expr << read_expression
+        #log expr[expr.length - 1]
+      end
+
+      skip_whitespace
+    end
+
+    gt = read_char # 5) Read '>'
+
+    if atom.length == 0 && expr.length == 0
+      Syntax::Form.new # empty Form
+    elsif atom.length != 0 && expr.length == 0
+      Syntax::Form.new(atom) # atom with no params
+    else
+      Syntax::Form.new(*[atom, *expr]) # atom and params

```

Comment:
redundant `*[]`
`Syntax::Form.new(atom, *expr)` should suffice

Reply:
I believe this code is correct.

Form.new is expecting a single array of values. Using `*[atom, expr]` instead returns an array with the atom in the first position and another array in the second position, instead of an array that includes atom followed by all of the values of expr.

Maybe I'm misunderstanding?

Reply:
I think levi is right.....
```rb
Syntax::Form.new(atom, *expr)
```

the outermost `*` and `[]` are just canceling each other out

Reply:
I stand corrected on this one as well.

Reply:
I wrote a unit test to prove it. ;)

```
def splat_elements(*elements)
  elements
end

def test_ruby_array(args, assert)

  atom = "1"
  expr = ["2", "3"]

  a1 = splat_elements(atom, *expr)
  a2 = splat_elements(*[atom, *expr])

  log a1.to_s
  log a2.to_s

  assert.equal! a1, a2
end
```

Reply:
This will be fixed in next commit.